free up dispatch key space (in C++)

bdhirsh opened the pull request:
Reland of https://github.com/pytorch/pytorch/pull/69633, after fixing some internal mobile failures.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #72828
* __->__ #72827

Differential Revision: [D34227616](https://our.internmc.facebook.com/intern/diff/D34227616/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D34227616/)!

bdhirsh committed:
free up dispatch key space (in C++)

Reland of D34034848

Differential Revision: [D34227616](https://our.internmc.facebook.com/intern/diff/D34227616/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D34227616/)!

[ghstack-poisoned]

bdhirsh committed:
Update on "free up dispatch key space (in C++)"


Reland of D34034848

Differential Revision: [D34227616](https://our.internmc.facebook.com/intern/diff/D34227616/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D34227616/)!

[ghstack-poisoned]

bdhirsh committed:
Update on "free up dispatch key space (in C++)"


Reland of D34034848

Differential Revision: [D34227616](https://our.internmc.facebook.com/intern/diff/D34227616/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D34227616/)!

[ghstack-poisoned]

bdhirsh committed:
Update on "free up dispatch key space (in C++)"


Reland of D34034848

Differential Revision: [D34227616](https://our.internmc.facebook.com/intern/diff/D34227616/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D34227616/)!

[ghstack-poisoned]

bdhirsh committed:
Update on "free up dispatch key space (in C++)"


Reland of D34034848

Differential Revision: [D34227616](https://our.internmc.facebook.com/intern/diff/D34227616/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D34227616/)!

[ghstack-poisoned]

bdhirsh committed:
Update on "free up dispatch key space (in C++)"


Reland of D34034848

Differential Revision: [D34227616](https://our.internmc.facebook.com/intern/diff/D34227616/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D34227616/)!

[ghstack-poisoned]

bdhirsh committed:
Update on "free up dispatch key space (in C++)"


Reland of D34034848

Differential Revision: [D34227616](https://our.internmc.facebook.com/intern/diff/D34227616/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D34227616/)!

[ghstack-poisoned]

pearu commented on aten/src/ATen/TensorSubclassLikeUtils.h:
```diff
@@ -28,8 +28,7 @@ constexpr auto kFunctorchWrappedTensors = DispatchKeySet({
-    DispatchKey::SparseCPU,
-    DispatchKey::SparseCUDA,
+    DispatchKey::Sparse,
     DispatchKey::SparseCsrCPU,
     DispatchKey::SparseCsrCUDA,
```
We'll likely need to detach CPU/CUDA from CSR dispatch keys as well, similar to `Sparse`:
```suggestion
    DispatchKey::SparseCsr,
```

bdhirsh replied:
I'll probably do that in a followup. For context: There actually isn't a hard requirement on which way to represent `SparseCsr` - it's more about whether or not we want to trade off dispatch key slots for space in the runtime operator table. We could either:

(1) Make it a "per-backend functionality key". There would be one dispatch key for `SparseCsr`, and any backend (CPU, CUDA, etc) could implement their own kernels for it separately. Advantage: we save some key space, and give every backend the flexibility to override its meaning (instead of being limited to just CPU and CUDA). Minor disadvantage: we blow up the size of the runtime operator table a bit, since we now need to store an extra (# backends) number of slots for SparseCsr, for every operator.

(2) Leave the existing `SparseCsrCPU/CUDA` keys, which saves some runtime operator table space.

It sounds like we should eventually do (1)

bdhirsh committed:
Update on "free up dispatch key space (in C++)"


Reland of D34034848

Differential Revision: [D34227616](https://our.internmc.facebook.com/intern/diff/D34227616/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D34227616/)!

[ghstack-poisoned]
